api/apps/api/src/modules/scenario-specification/scenario-specification.ts

```diff
@@ -0,0 +1,264 @@
+
+import { SpecificationOperation } from './specification-operation';
+import { FeaturesFound } from './features-found.event';
+import { CandidateSpecCreated } from './candidate-spec-created.event';
+import { SpecificationPublished } from './specification-published.event';
+import { SpecificationActivated } from './specification-actived.event';
+
+export class ScenarioSpecification extends AggregateRoot {
+  #currentActiveSpec?: Specification; // a one that was published and calculations made and is marked as active (effective)
+  #currentCandidateSpec?: Specification;
+
+  constructor(
+    public readonly scenarioId: string,
+    readonly currentActiveSpecification?: SpecificationSnapshotInput,
+    readonly specificationCandidate?: SpecificationSnapshotInput,
+  ) {
+    super();
+    this.#currentActiveSpec = currentActiveSpecification
+      ? Specification.from(currentActiveSpecification)
+      : undefined;
+    this.#currentCandidateSpec = specificationCandidate
+      ? Specification.from(specificationCandidate)
+      : undefined;
+  }
+
+  createCandidate(input: FeatureConfigInput[], draft: boolean) {
+    this.#currentCandidateSpec = Specification.new(input, draft);
+
+    const specSnapshot = this.#currentCandidateSpec.toSnapshot();
+    this.apply(
+      new CandidateSpecCreated(
+        specSnapshot.id,
+        specSnapshot.config.map((cfg) => ({
+          baseFeatureId: cfg.baseFeatureId,
+          againstFeatureId: cfg.againstFeatureId,
+          operation: cfg.operation,
+        })),
+      ),
+    );
+  }
+
+  /**
+   * consumes FeaturesFound
+   *
+   * may lead to activating Candidate Specification
+   * as long as all found features are already calculated
+   * and specification was marked as 'Create'
+   *
+   */
+  determineFeatures(features: FeaturesFound['spec']) {
+    this.#currentCandidateSpec?.determineFeatures(features);
+    this.publishCurrentCandidate();
+  }
+
+  /**
+   * may lead to activating the Candidate Spec
+   */
+  markFeaturesAsCalculated(features: string[]) {
+    if (this.#currentCandidateSpec?.markAsCalculated(features)) {
+      this.markCandidateAsActive();
+    }
+  }
+
+  getActiveSpecification = () => this.#currentActiveSpec?.toSnapshot();
+  getCandidateSpecification = () => this.#currentCandidateSpec?.toSnapshot();
+
+  private publishCurrentCandidate() {
+    if (!this.#currentCandidateSpec) {
+      return;
+    }
+    const outcome = this.#currentCandidateSpec.publish();
+
+    if (!outcome) {
+      return;
+    }
+
+    if (isLeft(outcome)) {
+      // if draft, its fine
+      // if created, we will wait anyway for all features
+      return;
+    }
+
+    if (outcome.right === SpecCanBePublished) {
+      this.apply(
+        new SpecificationPublished(
+          this.#currentCandidateSpec.id,
+          this.#currentCandidateSpec.relatedSubFeaturesRequiringCalculations(),
+        ),
+      );
+    }
+
+    if (outcome.right === SpecCanBeActivated) {
```

Comment:
isn't the check redundant? shouldn't we just hit `active` as we do with `publish`?